src/digests.rs

```diff
@@ -0,0 +1,211 @@
+//! # Multihash
+//!
+//! Implementation of [multihash](https://github.com/multiformats/multihash) in Rust.
+//!
+//! A `Multihash` is a structure that contains a hashing algorithm, plus some hashed data.
+//! A `MultihashRef` is the same as a `Multihash`, except that it doesn't own its data.
+//!
+
+use std::convert::TryFrom;
+
+use bytes::{BufMut, Bytes, BytesMut};
+use num_traits::cast::FromPrimitive;
+use unsigned_varint::{decode as varint_decode, encode as varint_encode};
+
+use crate::errors::{DecodeError, DecodeOwnedError};
+use crate::hashes::Code;
+
+/// Represents a valid multihash.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Multihash {
+    bytes: Bytes,
```

Comment:
Just curious: why are we using Bytes here? Bytes is good if you want to share multiple slices of some larger buffer without copying. But a multihash is typically self-contained. Using bytes here might have the advantage of avoiding a copy, but it carries the risk of dragging around much more memory than we actually need.

E.g. you get a large buffer from some network or disk interface as a Bytes, then take the first 32 bytes from the buffer into a multihash. This will then contain a pointer to the large buffer even if you keep the multihash for long and are no longer interested in the rest of the buffer.

Now, sharing is obviously important for a Multihash, but I would think that just using an Arc<[u8]> (not Arc<Vec<u8>> !) would be the best compromise.

E.g..

Reply:
Was also wondering about this but forgot to ask. Thanks @rklaehn for bringing it up

Reply:
I also wonder if dependency on `bytes` is really needed here. What was the original rationale (bytes vs. `Vec<_>`)?

For small amounts of data (like hashes) it might be always faster to just allocate a vector, like [sled might be doing in future](https://github.com/spacejam/sled/pull/942) than to do anything related to Arcs. Then again, perhaps my knowledge of `bytes` internals is outdated and I don't have time right now to understand them fully, could be that using `bytes` for single owner data is simple as `Vec<_>`.

Even if it was, as `bytes` is part of the async ecosystem it might be easier to get this crate stable throught the use of only mostly `std` components and to need to update this crate only when source specs or hash implementations evolve or more are needed.

Regarding @rklaehn from network usecase, I wonder if there is any benefit from actually constructing an owned type around the bytes, as the bytes are usable for verification purposes through the borrowed type, whatever it ends up looking like.

Also wondering if there should be a way to hash into a hopefully large enough `&mut [u8]` but perhaps that can be experimented outside of the crate if necessary. The same goes for updating on blocks. Perhaps this crate is at a quite difficult intersection with simplicity and different advanced IO usecases. I also wonder if only mapping from the spec codes to digest instances which can be used the by the caller in number of different ways would be enough.

Reply:
For the case where you don't want to own the thing, there is `MultihashRef`. And in general I must say fewer dependencies are always better, especially for crates that have not yet reached 1.0.

Reply:
> Even if it was, as bytes is part of the async ecosystem

Just to react to that: the `bytes` crate was introduced as a hack around the fact that you couldn't do anything combining asynchronous and borrowing. You couldn't for instance asynchronously write a buffer by reference.

This is no longer true in a post-async-await world, and I totally see `bytes` disappearing at some point.

Reply:
The background why `bytes` is used: https://github.com/multiformats/rust-multihash/pull/42.

Reply:
I would propose comment summarizing why `Bytes` are used, clearly without a good context it seems controversial.

Reply:
The discussion in https://github.com/multiformats/rust-multihash/pull/42 was comparing Bytes to Vec<u8>. I would love to see the tradeoff between Bytes and Arc<[u8]>. It might be that there are situations where Bytes comes up on top, but currently I don't see them.

Reply:
I've updated the Gist (https://gist.github.com/vmx/0dc0d981b522e2834084559faa0dcb56) and added the code from https://github.com/multiformats/rust-multihash/pull/47 as comparison. On my machine that new code is not slower than the bytes approach.

Reply:
re: sled use, I'm closing that linked issue because although it is faster to copy 2-4k than bump an Arc in some cases, it's NOT faster to allocate a vec first. I need to do more measuring to find out whether that third path between inlining and Arc<[u8]> actually exists.